File: doc-examples/example-java/src/test/java/example/RuntimeExpressionTest.java
package example;

import io.micronaut.el.CompiledELContext;
import io.micronaut.el.runtime.CompiledExpression;
import jakarta.el.ELContext;
import jakarta.el.ELManager;
import jakarta.el.ValueExpression;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

class RuntimeExpressionTest {

    @Test
    void parsesAnExpressionBuiltAtRuntime() {
        ELContext context = new CompiledELContext().setBean("book", new Book("Jakarta EL", "reference", 20d));
        String property = "category"; // <1>

        ValueExpression expression = ELManager.getExpressionFactory()
            .createValueExpression(context, "${book." + property + "}", String.class); // <2>

        assertFalse(expression instanceof CompiledExpression); // <3>
        assertEquals("reference", expression.getValue(context));
    }
}

File: doc-examples/example-kotlin/src/test/kotlin/example/RuntimeExpressionTest.kt
package example

import io.micronaut.el.CompiledELContext
import io.micronaut.el.runtime.CompiledExpression
import jakarta.el.ELManager
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test

class RuntimeExpressionTest {

    @Test
    fun parsesAnExpressionBuiltAtRuntime() {
        val context = CompiledELContext().setBean("book", Book("Jakarta EL", "reference", 20.0))
        val property = "category" // <1>

        val expression = ELManager.getExpressionFactory()
            .createValueExpression(context, "\${book.$property}", String::class.java) // <2>

        assertFalse(expression is CompiledExpression) // <3>
        assertEquals("reference", expression.getValue(context))
    }
}
